cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/class-sync.php

```diff
@@ -119,66 +163,654 @@ public function is_synced( $post_id ) {
+		 * Return a WP Error with reason why it can't be synced.
+		 *
+		 * @param int $attachment_id The attachment post ID.
+		 *
+		 * @return bool|\WP_Error
+		 */
+		return apply_filters( 'cloudinary_can_sync_asset', $can, $attachment_id, $type );
+	}
+
+	/**
+	 * Get the last version this asset was synced with.
+	 *
+	 * @param int $attachment_id The attachment ID.
+	 *
+	 * @return mixed
+	 */
+	public function get_sync_version( $attachment_id ) {
+		$version = $this->managers['media']->get_post_meta( $attachment_id, self::META_KEYS['plugin_version'], true );
+
+		return $version;
+	}
+
 	/**
 	 * Get the current sync signature of an asset.
 	 *
-	 * @param int $post_id The post ID.
+	 * @param int  $attachment_id The attachment ID.
+	 * @param bool $cached        Flag to specify if a cached signature is to be used or build a new one.
 	 *
 	 * @return array|bool
 	 */
-	public function get_signature( $post_id ) {
+	public function get_signature( $attachment_id, $cached = true ) {
 		static $signatures = array(); // Cache signatures already fetched.
 
-		$return = false;
-		if ( ! empty( $signatures[ $post_id ] ) ) {
-			$return = $signatures[ $post_id ];
+		$return = array();
+		if ( ! empty( $signatures[ $attachment_id ] ) && true === $cached ) {
+			$return = $signatures[ $attachment_id ];
 		} else {
-			$signature = $this->plugin->components['media']->get_post_meta( $post_id, self::META_KEYS['signature'], true );
-			if ( ! empty( $signature ) ) {
-				$base_signatures        = $this->generate_signature( $post_id );
-				$signatures[ $post_id ] = wp_parse_args( $signature, $base_signatures );
-				$return                 = $signatures[ $post_id ];
+			$signature = $this->managers['media']->get_post_meta( $attachment_id, self::META_KEYS['signature'], true );
+			if ( empty( $signature ) ) {
+				$signature = array();
 			}
+
+			// Remove any old or outdated signature items. against the expected.
+			$signature                    = array_intersect_key( $signature, $this->sync_types );
+			$signatures[ $attachment_id ] = $return;
+			$return                       = wp_parse_args( $signature, $this->sync_types );
 		}
 
 		return $return;
 	}
 
+	/**
+	 * Generate a new Public ID for an asset.
+	 *
+	 * @param int $attachment_id The attachment ID for the new public ID.
+	 *
+	 * @return string|null
+	 */
+	public function generate_public_id( $attachment_id ) {
+
+		$cld_folder = $this->managers['media']->get_cloudinary_folder();
+		$file       = get_attached_file( $attachment_id );
+		$file_info  = pathinfo( $file );
+		$public_id  = $cld_folder . $file_info['filename'];
+
+		return ltrim( $public_id, '/' );
+	}
+
+	/**
+	 * Maybe add a suffix to the public ID if it's not unique.
+	 *
+	 * @param string      $public_id     The public ID to maybe add a suffix.
+	 * @param int         $attachment_id The attachment ID.
+	 * @param string|null $suffix        The suffix to maybe add.
+	 *
+	 * @return string The public ID.
```

Comment:
```suggestion
	 * @return string|null The public ID.
```

Reply:
This will change with the tickets I'm working on this week